tests/test_core.py

```diff
@@ -8062,6 +8062,7 @@ def test_fpic_static(self):
     self.do_run_in_out_file_test('tests', 'core', 'test_hello_world.c')
 
   @node_pthreads
+  @unittest.skip("this test can not be run without a pool, see test_pthread_create_pool")
```

Comment:
But we do run this today, right?

Reply:
My understanding of `tests/core/pthread/create.cpp` is that it spawns 2 detached threads where the first thread blocks until the second one is spawned. Without setting `-sPTHREAD_POOL_SIZE=2` this can not be done properly, I think.

This test case was previously successful due to the `process.exit(status);` call. For example, testing this on master:
```diff
diff --git a/tests/core/pthread/create.cpp b/tests/core/pthread/create.cpp
index 1111111..2222222 100644
--- a/tests/core/pthread/create.cpp
+++ b/tests/core/pthread/create.cpp
@@ -23,7 +23,12 @@ void *ThreadMain(void *arg) {
     int last = sum.load();
     while (sum.load() == last) {}
   }
-  pthread_exit((void*)TOTAL);
+  EM_ASM({
+    PThread.threadExit($0);
+    //process.exit($0);
+    throw 'unwind';
+  }, TOTAL);
+  return nullptr;
 }
 
 pthread_t thread[NUM_THREADS];
```

Will also deadlock, uncommenting `process.exit($0);` makes the test case "pass".

Reply:
I see.  So why not add PTHREAD_POOL_SIZE to this test rather than disabling it?

Reply:
I could add `PTHREAD_POOL_SIZE`, but then the test case is essentially the same as `test_pthread_create_pool`, see:
https://github.com/emscripten-core/emscripten/blob/62007d1adbdccc56c3e1ea9e7bb1aecd6e89f1ea/tests/test_core.py#L8066-L8071

Reply:
I see.   It seems like the code is supposed to honor `ALLOW_SYNC` and when that is not defined it should not depend on synchronous thread creation.  Maybe we need to fix the code?

Reply:
You say "is that it spawns 2 detached threads where the first thread blocks until the second one is spawned"... I'm a little confused.  I don't see any blocking happening the create.cpp code.

Reply:
Oh, I thought it was blocking on:
https://github.com/emscripten-core/emscripten/blob/b97bbc55a177a532c9d689e4d764c0761c1a7923/tests/core/pthread/create.cpp#L22-L24

fwiw, here's a native program: https://gist.github.com/kleisauke/232be709d82b6aaecfa0f28a8be27fe3

(with that program I see that it runs 2 threads next to each other, I suppose that isn't supported unless you set `PTHREAD_POOL_SIZE` to a number equal to or greater than 2(?)).

Reply:
thats not quite true.   Without `PTHREAD_POOL_SIZE` emscripten needs to start new workers for each thread, which means you need to return to event loop before the thread will actually start.   This means what you can't do is not return from main.. this code in create.cpp as two modes: if ALLOW_SYNC is set it will busy wait in main (and not return), if ALLOW_SYNC is not set it will run an event loop by calling emscripten_set_main_loop (which returns from main).

Reply:
another way of putting: the thread pool is only needed if you need to do synchronous thread creation.   At least IIUC.

Reply:
Ah yes, you're right. This test case isn't blocking, sorry for the noise. It looks like the Node workers are not teared down when the main loop is canceled. I could fix this with explicitly calling `PThread.terminateAllThreads();` on exit, see for example: https://github.com/kleisauke/emscripten/commit/6141e7e3105cc9a8dc15a9e6b887451501d9b3a1.

I think this behavior is being investigated within issue #12801.